### ISSUE TITLE
[LAA] LoopAccessAnalysis depends on order of blocks in loop and may give false-positive safety answers

### ISSUE THREAD

**Issue by max-quazan** [llvm/llvm-project#56672]
I am not sure if this particular thing is a bug, but looks *extremely* fishy.

Link 1: https://godbolt.org/z/1bfKdadx4
Link 2: https://godbolt.org/z/4hq1931ex

Test:
```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define void @test()  {
bb:
  br label %bb4

bb1:                                              ; preds = %bb4
  br label %bb2

bb2:                                              ; preds = %bb1
  store i32 1, i32 addrspace(1)* %tmp7, align 4
  %tmp = add nuw i64 %tmp5, 1
  %tmp3 = icmp ult i64 %tmp, undef
  br i1 %tmp3, label %bb4, label %bb11

bb4:                                              ; preds = %bb2, %bb
  %tmp5 = phi i64 [ %tmp, %bb2 ], [ 16, %bb ]
  %tmp6 = phi i64 [ %tmp5, %bb2 ], [ 15, %bb ]
  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
  %tmp8 = load i32, i32 addrspace(1)* %tmp7, align 4
  %tmp9 = add i32 %tmp8, -5
  store i32 %tmp9, i32 addrspace(1)* %tmp7, align 4
  br i1 false, label %bb10, label %bb1

bb10:                                             ; preds = %bb4
  unreachable

bb11:                                             ; preds = %bb2
  ret void
}
```

Two run commands:
```shell
opt -passes='loop(loop-rotate),loop-vectorize' -S
opt -passes='loop(loop-rotate),invalidate<loops>,loop-vectorize' -S
```

In the former case, loop does vectorize. In the later case, it doesn't. For us, the vectorization is buggy (miscompile), but I could not figure out how to reduce test to demonstrate the miscompile yet.

The suspicion is that some analysis is in broken state after loop-rotate, and LV takes wrong decision basing on it.

**Comment by max-quazan** [comment]
Vectorized IR as is now:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define void @test() {
bb:
  %tmp81 = load i32, i32 addrspace(1)* getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), align 4
  %tmp92 = add i32 %tmp81, -5
  store i32 %tmp92, i32 addrspace(1)* getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), align 4
  %umax5 = call i64 @llvm.umax.i64(i64 undef, i64 17)
  %0 = add i64 %umax5, -16
  %min.iters.check = icmp ule i64 %0, 16
  br i1 %min.iters.check, label %scalar.ph, label %vector.scevcheck

vector.scevcheck:                                 ; preds = %bb
  %umax = call i64 @llvm.umax.i64(i64 undef, i64 17)
  %1 = add i64 %umax, -17
  %mul = call { i64, i1 } @llvm.umul.with.overflow.i64(i64 4, i64 %1)
  %mul.result = extractvalue { i64, i1 } %mul, 0
  %mul.overflow = extractvalue { i64, i1 } %mul, 1
  %2 = sub i64 0, %mul.result
  %3 = getelementptr i8, i8 addrspace(1)* bitcast (i32 addrspace(1)* getelementptr (i32, i32 addrspace(1)* null, i64 16) to i8 addrspace(1)*), i64 %mul.result
  %4 = icmp ult i8 addrspace(1)* %3, bitcast (i32 addrspace(1)* getelementptr (i32, i32 addrspace(1)* null, i64 16) to i8 addrspace(1)*)
  %5 = or i1 %4, %mul.overflow
  br i1 %5, label %scalar.ph, label %vector.ph

vector.ph:                                        ; preds = %vector.scevcheck
  %n.mod.vf = urem i64 %0, 4
  %6 = icmp eq i64 %n.mod.vf, 0
  %7 = select i1 %6, i64 4, i64 %n.mod.vf
  %n.vec = sub i64 %0, %7
  %ind.end = getelementptr i32, i32 addrspace(1)* getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), i64 %n.vec
  %ind.end6 = add i64 16, %n.vec
  br label %vector.body

vector.body:                                      ; preds = %vector.body, %vector.ph
  %index = phi i64 [ 0, %vector.ph ], [ %index.next, %vector.body ]
  %vec.ind = phi <4 x i64> [ <i64 16, i64 17, i64 18, i64 19>, %vector.ph ], [ %vec.ind.next, %vector.body ]
  %8 = add i64 %index, 0
  %next.gep = getelementptr i32, i32 addrspace(1)* getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), i64 %8
  %9 = getelementptr i32, i32 addrspace(1)* %next.gep, i32 0
  %10 = bitcast i32 addrspace(1)* %9 to <4 x i32> addrspace(1)*
  store <4 x i32> <i32 1, i32 1, i32 1, i32 1>, <4 x i32> addrspace(1)* %10, align 4
  %11 = add nuw <4 x i64> %vec.ind, <i64 1, i64 1, i64 1, i64 1>
  %12 = extractelement <4 x i64> %11, i32 0
  %13 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %12
  %14 = getelementptr inbounds i32, i32 addrspace(1)* %13, i32 0
  %15 = bitcast i32 addrspace(1)* %14 to <4 x i32> addrspace(1)*
  %wide.load = load <4 x i32>, <4 x i32> addrspace(1)* %15, align 4
  %16 = add <4 x i32> %wide.load, <i32 -5, i32 -5, i32 -5, i32 -5>
  %17 = bitcast i32 addrspace(1)* %14 to <4 x i32> addrspace(1)*
  store <4 x i32> %16, <4 x i32> addrspace(1)* %17, align 4
  %index.next = add nuw i64 %index, 4
  %vec.ind.next = add <4 x i64> %vec.ind, <i64 4, i64 4, i64 4, i64 4>
  %18 = icmp eq i64 %index.next, %n.vec
  br i1 %18, label %middle.block, label %vector.body, !llvm.loop !0

middle.block:                                     ; preds = %vector.body
  br label %scalar.ph

scalar.ph:                                        ; preds = %vector.scevcheck, %bb, %middle.block
  %bc.resume.val = phi i32 addrspace(1)* [ %ind.end, %middle.block ], [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %vector.scevcheck ]
  %bc.resume.val7 = phi i64 [ %ind.end6, %middle.block ], [ 16, %bb ], [ 16, %vector.scevcheck ]
  br label %bb1

bb1:                                              ; preds = %scalar.ph, %bb4
  %tmp74 = phi i32 addrspace(1)* [ %bc.resume.val, %scalar.ph ], [ %tmp7, %bb4 ]
  %tmp53 = phi i64 [ %bc.resume.val7, %scalar.ph ], [ %tmp5, %bb4 ]
  br label %bb2

bb2:                                              ; preds = %bb1
  store i32 1, i32 addrspace(1)* %tmp74, align 4
  %tmp = add nuw i64 %tmp53, 1
  %tmp3 = icmp ult i64 %tmp, undef
  br i1 %tmp3, label %bb4, label %bb11

bb4:                                              ; preds = %bb2
  %tmp5 = phi i64 [ %tmp, %bb2 ]
  %tmp6 = phi i64 [ %tmp53, %bb2 ]
  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
  %tmp8 = load i32, i32 addrspace(1)* %tmp7, align 4
  %tmp9 = add i32 %tmp8, -5
  store i32 %tmp9, i32 addrspace(1)* %tmp7, align 4
  br i1 false, label %bb10, label %bb1, !llvm.loop !2

bb10:                                             ; preds = %bb4
  unreachable

bb11:                                             ; preds = %bb2
  ret void
}

declare i64 @llvm.umax.i64(i64, i64) #0

declare { i64, i1 } @llvm.umul.with.overflow.i64(i64, i64) #0

attributes #0 = { nocallback nofree nosync nounwind readnone speculatable willreturn }

!0 = distinct !{!0, !1}
!1 = !{!"llvm.loop.isvectorized", i32 1}
!2 = distinct !{!2, !1}
```

Note that opt v14.0 did not vectorize this. The issue must be recent, but at least 1 month old.

**Comment by max-quazan** [comment]
Our triage found that this test started vectorizing with
```
commit d663166acbe59e4f17e406302552a1a9005204e8
Author: Simon Pilgrim <llvm-dev@redking.me.uk>
Date: Fri Mar 25 11:09:40 2022 +0000

[CostModel][X86] Reduce cost of v2i64 icmp base cost on SSE2 targets

Based off the script from D103695, we were exaggerating the cost of the v2i64 comparison expansion using instruction count instead of effective throu
```

I don't think this patch might be of any harm. It just means that the issue is at least as old as this one.

**Comment by max-quazan** [comment]
Difference when forgetting/not forgetting the loop is
```
LV: We can vectorize this loop!
```
vs
```
LV: Can't vectorize due to memory conflicts
```

**Comment by max-quazan** [comment]
Looks like loop access analysis is broken after loop rotate. Here is output with `-debug-only="loop-vectorize,loop-accesses"` with and without "invalidate".

When invalidating:
```
LV: Checking a loop in 'test' from ./upstream.ll
LV: Loop hints: force=? width=0 interleave=0
LV: Found a loop: bb1
LV: Found an induction variable.
LV: Found an induction variable.
LV: Did not find one integer induction var.
LAA: Found a loop in test: bb1
LAA: Processing memory accesses...
  AST: Alias Set Tracker: 1 alias sets for 2 pointer values.
  AliasSet[0x61fac30, 2] may alias, No access Pointers: (i32 addrspace(1)* %tmp74, unknown before-or-after), (i32 addrspace(1)* %tmp7, unknown before-or-after)

LAA:   Accesses(3):
	  %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ] (write)
	  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5 (write)
	  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5 (read)
Underlying objects for pointer   %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ]
  i32 addrspace(1)* null
Underlying objects for pointer   %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
  i32 addrspace(1)* null
Underlying objects for pointer   %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
  i32 addrspace(1)* null
LAA: Found a runtime check ptr:  %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ]
LAA: Found a runtime check ptr:  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
LAA: We need to do 0 pointer comparisons.
LAA: May be able to perform a memory runtime check if needed.
LAA: Checking memory dependencies
LAA: Src Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>Sink Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>(Induction step: 1)
LAA: Distance for   %tmp8 = load i32, i32 addrspace(1)* %tmp7, align 4 to   store i32 %tmp9, i32 addrspace(1)* %tmp7, align 4: 0
LAA: Pointer may wrap in the address space:
LAA:   Pointer:   %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ]
LAA:   SCEV: {(64 + null)<nuw><nsw>,+,4}<%bb1>
LAA:   Added an overflow assumption
LAA: Src Scev: {(64 + null)<nuw><nsw>,+,4}<%bb1>Sink Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>(Induction step: 1)
LAA: Distance for   store i32 1, i32 addrspace(1)* %tmp74, align 4 to   %tmp8 = load i32, i32 addrspace(1)* %tmp7, align 4: 4
LAA: Failure because of positive distance 4
LAA: Src Scev: {(64 + null)<nuw><nsw>,+,4}<%bb1>Sink Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>(Induction step: 1)
LAA: Distance for   store i32 1, i32 addrspace(1)* %tmp74, align 4 to   store i32 %tmp9, i32 addrspace(1)* %tmp7, align 4: 4
LAA: Failure because of positive distance 4
Total Dependences: 3
LAA: unsafe dependent memory operations in loop
LV: Can't vectorize due to memory conflicts
```
When not invalidating:
```
LV: Checking a loop in 'test' from ./upstream.ll
LV: Loop hints: force=? width=0 interleave=0
LV: Found a loop: bb1
LV: Found an induction variable.
LV: Found an induction variable.
LV: Did not find one integer induction var.
LAA: Found a loop in test: bb1
LAA: Processing memory accesses...
  AST: Alias Set Tracker: 1 alias sets for 2 pointer values.
  AliasSet[0x615f160, 2] may alias, No access Pointers: (i32 addrspace(1)* %tmp7, unknown before-or-after), (i32 addrspace(1)* %tmp74, unknown before-or-after)

LAA:   Accesses(3):
	  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5 (write)
	  %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ] (write)
	  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5 (read)
Underlying objects for pointer   %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
  i32 addrspace(1)* null
Underlying objects for pointer   %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
  i32 addrspace(1)* null
Underlying objects for pointer   %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ]
  i32 addrspace(1)* null
LAA: Found a runtime check ptr:  %tmp7 = getelementptr inbounds i32, i32 addrspace(1)* null, i64 %tmp5
LAA: Found a runtime check ptr:  %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ]
LAA: We need to do 0 pointer comparisons.
LAA: May be able to perform a memory runtime check if needed.
LAA: Checking memory dependencies
LAA: Pointer may wrap in the address space:
LAA:   Pointer:   %tmp74 = phi i32 addrspace(1)* [ getelementptr inbounds (i32, i32 addrspace(1)* null, i64 16), %bb ], [ %tmp7, %bb4 ]
LAA:   SCEV: {(64 + null)<nuw><nsw>,+,4}<%bb1>
LAA:   Added an overflow assumption
LAA: Src Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>Sink Scev: {(64 + null)<nuw><nsw>,+,4}<%bb1>(Induction step: 1)
LAA: Distance for   %tmp8 = load i32, i32 addrspace(1)* %tmp7, align 4 to   store i32 1, i32 addrspace(1)* %tmp74, align 4: -4
LAA: Dependence is negative
LAA: Src Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>Sink Scev: {(64 + null)<nuw><nsw>,+,4}<%bb1>(Induction step: 1)
LAA: Distance for   store i32 %tmp9, i32 addrspace(1)* %tmp7, align 4 to   store i32 1, i32 addrspace(1)* %tmp74, align 4: -4
LAA: Dependence is negative
LAA: Src Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>Sink Scev: {(68 + null)<nuw><nsw>,+,4}<%bb1>(Induction step: 1)
LAA: Distance for   %tmp8 = load i32, i32 addrspace(1)* %tmp7, align 4 to   store i32 %tmp9, i32 addrspace(1)* %tmp7, align 4: 0
Total Dependences: 3
LAA: No unsafe dependent memory operations in loop.  We don't need runtime memory checks.
LV: We can vectorize this loop!
```

**Comment by max-quazan** [comment]
Loops like LAA's algorithm depends on block order in loop info. This order is arbitrary and changes after invalidation.

**Comment by max-quazan** [comment]
https://reviews.llvm.org/D130482

**Comment by max-quazan** [comment]
```
commit 2d1c6e0b4418e92df286cba428a3e4cf56d7aa2f
Author: Max Kazantsev <mkazantsev@azul.com>
Date:   Thu Jul 28 13:27:19 2022 +0700

    [LAA] Remove block order sensitivity in LAA algorithm. PR56672

    As test in PR56672 shows, LAA produces different results which lead to either
    positive or negative vectorization decisions depending on the order of blocks
    in loop. The exact reason of this is not clear to me, however this makes investigation
    of related bugs extremely complex.

    Current order of blocks in the loop is arbitrary. It may change, for example, if loop
    info analysis is dropped and recomputed. Seems that it interferes with LAA's logic.
    This patch chooses fixed traversal order of blocks in loops, making it RPOT.

    Note: this is *not* a fix for bug with incorrect analysis result. It just makes
    the answer more robust to make the investigation easier.

    Differential Revision: https://reviews.llvm.org/D130482
    Reviewed By: aeubanks, fhahn
```
